mail_embed_image/models/ir_mail_server.py

```diff
@@ -0,0 +1,132 @@
+            email_bcc=None,
+            reply_to=None,
+            attachments=None,
+            message_id=None,
+            references=None,
+            object_id=None,
+            subtype='plain',
+            headers=None,
+            body_alternative=None,
+            subtype_alternative='plain'):
+        result = super(IrMailServer, self).build_email(
+            email_from=email_from,
+            email_to=email_to,
+            subject=subject,
+            body=body,
+            email_cc=email_cc,
+            email_bcc=email_bcc,
+            reply_to=reply_to,
+            attachments=attachments,
+            message_id=message_id,
+            references=references,
+            object_id=object_id,
+            subtype=subtype,
+            headers=headers,
+            body_alternative=body_alternative,
+            subtype_alternative=subtype_alternative,
+        )
+        return self._build_email_replace_img_src(result)
+
+    def _build_email_replace_img_src(self, email):
+        """ Given a message, find it's img tags and if they
+        are URLs, replace them with cids.
+        """
+        base_url = self.env['ir.config_parameter'].get_param(
+            'web.base.url')
+        for part in email.walk():
+            if part.get_content_maintype() == 'text':
+                body = part.get_payload(decode=True)
+                root = fromstring(body)
+                for img in root.xpath(
+                        "//img[starts-with(@src, '%s/web/image')]"
+                        "|"
+                        "//img[starts-with(@src, '/web/image')]" % (
+                            base_url, )):
+                    response = self._get_image(
+                        base_url,
+                        img.get('src').replace(base_url, ''))
+                    if not response:
+                        logger.warning('Could not get %s', img.get('src'))
+                        continue
+                    cid = uuid.uuid4().hex
+                    filename_rfc2047 = encode_header_param(cid)
+                    filepart = MIMEImage(response)
+                    filepart.set_param('name', filename_rfc2047)
+                    filepart.add_header(
+                        'Content-Disposition',
+                        'inline',
+                        cid=cid,
+                        filename=filename_rfc2047,
+                    )
+                    filepart.set_payload(response)
+                    Encoders.encode_base64(filepart)
+                    # attach the image into the email as attachment
+                    email.attach(filepart)
+                    img.set('src', 'cid:%s' % (cid))
+                part.set_payload(b64encode(tostring(root)))
+        return email
+
+    def _get_image(self, base_url, src):
+        """ This function emulates the operations of
+        odoo.addons.web.controllers.main.Binary.content_image() because
+        we cannot call this from here
```

Comment:
this was what I expected you to figure out. To be sure that I don't ask the impossible, here a version that runs in python_odoo:

```python
from odoo import http
from werkzeug.test import EnvironBuilder
from mock import Mock, patch

builder = EnvironBuilder(path='/web/image/1')
env = builder.get_environ()
endpoint, arguments = http.routing_map(
    session.env.registry._init_modules, 
    False, 
    converters=session.env['ir.http']._get_converters()
).bind_to_environ(
    env
).match(return_rule=False)

result_dict = {}
with patch(
    'odoo.addons.web.controllers.main.request', 
    env=session.env, 
    registry=session.env.registry,
    make_response = lambda x, y: result_dict.update(
        data=x,
        headers=y,
    ) or Mock(),
) as mock_request, patch(
    'odoo.addons.website.models.ir_http.request',
    env=session.env,
):
    result = endpoint.method(**arguments)
# without the mocking, result contains what you need
print len(result_dict['data'])
print result_dict['headers']
```

The problem with your code is twofold: The code duplication is already bad, but it's worse that you miss possible overrides of the controller that might change the image in whatever ways we can't foresee.

Adapt the above to your situation, and research why and how this works. Add comments with explanations and links to code/documentation that prove you did so. The mocking is ugly, better detect if `odoo.http.request` is bound, and if not, bind it to some fake request and unbind afterwards.